py/obj: Add requirement for 4-byte alignment

MikeWang000000 opened the pull request:
<!-- Thanks for submitting a Pull Request! We appreciate you spending the
     time to improve MicroPython. Please provide enough information so that
     others can review your Pull Request.

     Before submitting, please read:
     https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md
     https://github.com/micropython/micropython/wiki/ContributorGuidelines

     Please check any CI failures that appear after your Pull Request is opened.
-->

### Summary

<del>This change checks `__BIGGEST_ALIGNMENT__` and produces a compile-time error if the maximum alignment is less than 4 bytes. On m68k platforms, an additional message advises using `-malign-int` to enable 4-byte alignment.</del>

<del>This ensures correct operation of object tagging schemes that require at least 4-byte alignment.</del>

Add a compile-time static assertion to ensure uPy objects are aligned to at least a 4-byte boundary.

Change the MICROPY_OBJ_BASE_ALIGNMENT macro to a numeric value, making it easier to define via CFLAGS in Makefiles.

Reference: https://wiki.debian.org/M68k/Alignment
> On m68k, the default alignment has been traditionally 2 bytes instead of 4 bytes. This originates in the design of the original 68000 CPU which used an external 16-bit data bus such that data words with a length of 1 or 2 bytes could be handled by the data bus in one clock cycle instead of two. Starting with the 68020 CPU, the external data bus was extended to 32 bits and the previously used 2 bytes default alignment was no longer necessary to optimize performance of the data bus.

Related issue: #18183

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->


### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->

Tested on Linux, m68k-linux-gnu-gcc (Debian 14.2.0-19) 14.2.0

Successful build, REPL launched without errors.

dpgeorge commented:
Thanks for the contribution.

This check seems quite specific to m68k.  Can it be done in a more general way?

MikeWang000000 committed:
py/obj: Require 4-byte alignment for objects.

Add a compile-time static assertion to ensure uPy objects are aligned
to at least a 4-byte boundary.

Change the MICROPY_OBJ_BASE_ALIGNMENT macro to a numeric value,
making it easier to define via CFLAGS in Makefiles.

Signed-off-by: Mike Wang <mikewang000000@gmail.com>

MikeWang000000 commented:
Yes, `-malign-int` is a specific GCC option for m68k targets only. I updated the code to use `MICROPY_OBJ_BASE_ALIGNMENT` macro in a more general way.

---

Updated changes:

Add a compile-time static assertion to ensure uPy objects are aligned to at least a 4-byte boundary.

Change the MICROPY_OBJ_BASE_ALIGNMENT macro to a numeric value, making it easier to define via CFLAGS in Makefiles.

MikeWang000000 commented:
I didn't expect that all MSVC builds failed due to the new added static assert.
Interesting...

AJMansfield commented:
> I didn't expect that all MSVC builds failed due to the new added static assert. Interesting...

I'm not surprised --- there's actually like a dozen cases you need to cover if you want anything to do with alignment in C to be totally portable across all the compilers MicroPython targets.

I have a PR that adds proper portable macros for alignment, though, actually just as of this week:

-  #18148

Changing `MICROPY_OBJ_BASE_ALIGNMENT` to just be a number rather than an attribute sounds like a good idea to me, though --- I'd suggest rebasing downstream of that and changing the declaration in obj.h to use the `MP_ALIGNAS` macro it defines.

MikeWang000000 commented:
> I'd suggest rebasing downstream of that and changing the declaration in obj.h to use the MP_ALIGNAS macro it defines.

Sure! I will update the code when PR #18145 and your PR #18148 get merged. Thanks for pointing this out!
